### PR TITLE
Ensure backend copy

### LINKED ISSUE #80: Add an E2E test workflow
There are [Go integration tests](https://github.com/TakeScoop/terraform-cloud-workspace-action/blob/master/internal/action/main_test.go), but we should also have a full integration test of the Action:

```yml
uses: ./
with:
  terraform_organization: takescoop-oss
```

We could probably be clever here and supply a local backend and then invoke `terraform destroy` in a later step to remove the associated infrastructure

### PR DESCRIPTION
**Pull request by ryanwholey**
Fixes a bug where we are not writing the final state to the passed backend.

This fix inclued two changes, I think which makes the code easier to reason about.

Changes:

1) **before**: always copy to empty backend and only copy back to the provided backend at the end if apply.
**now**: switch to empty backend after init only if !apply. It was a bit confusing wondering whether the backend was empty or not at which point. This way theres one logic branch where, after loading state from the passed backend, we either switch or we do not.

1) **before**: was stingy about ensuring when state was copied, writing the import files locally without copying state over first to save some time (harder to reason about what to do when)
**now** use tf init exclusively to ensure that state is predictably transferred between backends, remove wrapper `CopyStateToBackend` function in favor of TerraformInit with a comment, as it's more direct to what's really happening.

Added an E2E test and was aiming to use terraform destroy, but I think we get a little more coverage by intentionally deleting the resources we created. Open to improvements.

closes #80

### CODE REVIEW

**Comment by ryanwholey** [comment]
Along with the e2e, tested in [takescoop/test-workspace](https://github.com/TakeScoop/test-workspace),

[test-workspace#16](https://github.com/TakeScoop/test-workspace/pull/16/files) updates some variables, the [plan](https://github.com/TakeScoop/test-workspace/runs/3592440542?check_suite_focus=true) accurately reflects them, then on apply, the we see the [same plan](https://github.com/TakeScoop/test-workspace/runs/3592445294?check_suite_focus=true) and success.

Also confirmed that we're adding s3 object versions to the state file there, where previously we were not

**Comment by ryanwholey** [comment]
Here's an interesting one, when updating a var from sensitive to non-sensitive, we get a failure without intervention (was curious what would happen)
https://github.com/TakeScoop/test-workspace/runs/3592680264

I don't think this is a bad outcome because this is intentionally like this to prevent leaking secrets, but interesting to think about...

**Comment by bendrucker** [comment]
> Here's an interesting one, when updating a var from sensitive to non-sensitive

This should only apply if the value is also unchanged. But yeah, small enough corner case that users should probably just deal with it.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+                category: terraform
+              - key: environment
+                value: production
+                category: terraform
+          workspaces: |-
+            - staging
+            - production
+          backend_config: |-
+            local:
+              path: /github/workspace/terraform.tfstate
```
Comment (bendrucker) [inline .github/workflows/test.yml:75]: Can `github.workspace` be used here?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+          workspace_variables: |-
+            staging:
+              - key: environment
+                value: staging
+                category: terraform
+              - key: environment
+                value: production
+                category: terraform
+          workspaces: |-
+            - staging
```
Comment (bendrucker) [inline .github/workflows/test.yml:71]: Use fake values here like alpha, beta

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+        with:
+          runner_terraform_version: '1.0.5'
+          terraform_organization: takescoop-oss
+          terraform_token: ${{ secrets.TF_TOKEN_OSS }}
+          name: ${{ steps.e2e-ws-name.outputs.name }}
+          tags: |-
+            - terraform:true
+          workspace_tags: |-
+            staging:
+              - environment:staging
```
Comment (bendrucker) [inline .github/workflows/test.yml:54]: Avoid spreading `environment` around beyond a usage example

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+            - production
+          backend_config: |-
+            local:
+              path: /github/workspace/terraform.tfstate
+      - name: "E2E: Assert & Cleanup"
+        env:
+          TF_TOKEN: ${{ secrets.TF_TOKEN_OSS }}
+        shell: bash
+        run: |
+          set -e
```
Comment (bendrucker) [inline .github/workflows/test.yml:81]: This probably happens by default

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+
+          stat terraform.tfstate > /dev/null && echo "Success: state file present"
+
+          for ws in staging production ; do
+            curl -fsS \
```
Comment (bendrucker) [inline .github/workflows/test.yml:86]: Full flags

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+      - name: "E2E: Assert & Cleanup"
+        env:
+          TF_TOKEN: ${{ secrets.TF_TOKEN_OSS }}
+        shell: bash
+        run: |
+          set -e
+
+          stat terraform.tfstate > /dev/null && echo "Success: state file present"
+
+          for ws in staging production ; do
```
Comment (bendrucker) [inline .github/workflows/test.yml:85]: Make this an output for reuse in both places. Duplication in assertions can be good for readability but cleanup functions should generally share a reference to guarantee resource removal.

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+              - key: environment
+                value: production
+                category: terraform
+          workspaces: |-
+            - staging
+            - production
+          backend_config: |-
+            local:
+              path: /github/workspace/terraform.tfstate
+      - name: "E2E: Assert & Cleanup"
```
Comment (bendrucker) [inline .github/workflows/test.yml:76]: Split assertions from cleanup. Assertions should run on success (default). Cleanup should run always. A run could fail on related workspace resources but the workspaces still require removal.

File: .github/workflows/test.yml (hunk 6 above)
Comment (bendrucker) [reply .github/workflows/test.yml:85]: Also put each ws in a group. Log pretty verbosely and let the groups hide most of the details.

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -15,23 +15,78 @@ jobs:
...
+          set -e
+
+          stat terraform.tfstate > /dev/null && echo "Success: state file present"
+
+          for ws in staging production ; do
+            curl -fsS \
+              --header "Authorization: Bearer $TF_TOKEN" \
+              --header "Content-Type: application/vnd.api+json" \
+              --request DELETE \
+              "https://app.terraform.io/api/v2/organizations/takescoop-oss/workspaces/${{ steps.e2e-ws-name.outputs.name }}-$ws" > /dev/null
```
Comment (bendrucker) [inline .github/workflows/test.yml:90]: Make sure this doesn't swallow an error response that would normally be printed. Not sure if `--fail` causes output to be written to stderr instead. With groups, don't hesitate to print stuff.

**Review by bendrucker — APPROVED** [review]
Nice, agree with the approach, you can merge and release this with a follow up if you want
